### PR TITLE
Using ensime for autocompletion and type hints

### LINKED ISSUE #224: Eat characters when autocompleting
```
List(1,2,3).
           ^ autocomplete
```

then typing `rev` should filter the autocompletion box    (reverse, ...)

### PR DESCRIPTION
**Pull request by MasseGuillaume**
* fix #224: eat characters
* reuse Sbt instance in EnsimeActor
* change css for hints

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -189,7 +188,7 @@ lazy val sbtRunner = project
         val logbackConfDestination = "/root/logback.xml"
 
         new Dockerfile {
-          from("scalacenter/scastie-docker-sbt:0.0.24")
+          from("scalacenter/scastie-docker-sbt:0.0.25")
```
Comment (MasseGuillaume) [inline build.sbt]: we need to bump the docker image if anything changes

File: client/src/main/scala/com.olegych.scastie.client/components/Editor.scala [hunk 2]
```diff
@@ -107,11 +116,29 @@ object Editor {
...
   }
 
+  /**
+   *    +------------------------+-------------+
+   *    v                        |             |
+   *   Idle --> Requested --> Active <--> NeedRender
+   *    ^           |
+   *    +-----------+
+   *   only if exactly one
+   *   completion returned
```
Comment (MasseGuillaume) [inline client/src/main/scala/com.olegych.scastie.client/components/Editor.scala]: really cool!

File: sbt-runner/src/main/scala/com.olegych.scastie.sbt/EnsimeActor.scala [hunk 3]
```diff
@@ -1,87 +1,87 @@
...
-import JerkyFormats._
+
+import org.slf4j.LoggerFactory
+
+import java.io.{BufferedReader, File, InputStream, InputStreamReader}
+import java.nio.file.{Files, Path}
+
+import scala.concurrent.duration._
+import scala.concurrent.Future
+import scala.concurrent.ExecutionContext.Implicits.global
```
Comment (MasseGuillaume) [inline sbt-runner/src/main/scala/com.olegych.scastie.sbt/EnsimeActor.scala]: you can 'import system.dispatcher'

File: sbt-runner/src/main/scala/com.olegych.scastie.sbt/EnsimeActor.scala [hunk 4]
```diff
@@ -1,87 +1,87 @@
...
-  private val sbt = new Sbt(
-    defaultConfig,
-    rootDir,
-    secretSbtConfigExtra = sbtConfigExtra,
-    secretSbtPluginsConfigExtra = sbtPluginsConfigExtra
-  )
-
-  private val httpPortFile = ensimeCacheDir.resolve("http")
-
   private var ensimeProcess: Process = _
```
Comment (MasseGuillaume) [inline sbt-runner/src/main/scala/com.olegych.scastie.sbt/EnsimeActor.scala]: Option[Process]

File: sbt-runner/src/main/scala/com.olegych.scastie.sbt/EnsimeActor.scala [hunk 5]
```diff
@@ -1,87 +1,87 @@
...
-  private val ensimeCacheDir = rootDir.resolve(".ensime_cache")
-  Files.createDirectories(ensimeCacheDir)
-
-  private val sbtConfigExtra = s"""
-                                  |// this is where the ensime-server snapshots are hosted
-                                  |resolvers += Resolver.sonatypeRepo("snapshots")
-                                  |libraryDependencies += "org.ensime" %% "ensime" % "$ensimeVersion"
-                                  |""".stripMargin
-  private val sbtPluginsConfigExtra =
-    s"""addSbtPlugin("org.ensime" % "sbt-ensime" % "1.12.11")""".stripMargin
```
Comment (MasseGuillaume) [inline sbt-runner/src/main/scala/com.olegych.scastie.sbt/EnsimeActor.scala]: are you sure this is not necessary?

**Comment by MasseGuillaume** [comment]
when I do `sbtRunner/reStart`, I get:

```
[info] Application sbtRunner not yet started
[info] Starting application sbtRunner in the background ...
sbtRunner Starting com.olegych.scastie.sbt.SbtMain.main()
[success] Total time: 36 s, completed Jul 12, 2017 11:52:25 AM
> sbtRunner 11:52:26.761 INFO  SbtMain:19 | Starting sbtRunner pid: 11122
sbtRunner [INFO] [07/12/2017 11:52:27.511] [main] [akka.remote.Remoting] Starting remoting
sbtRunner [INFO] [07/12/2017 11:52:27.690] [main] [akka.remote.Remoting] Remoting started; listening on addresses :[akka.tcp://SbtRemote@127.0.0.1:5150]
sbtRunner [INFO] [07/12/2017 11:52:27.693] [main] [akka.remote.Remoting] Remoting now listens on addresses: [akka.tcp://SbtRemote@127.0.0.1:5150]
sbtRunner 11:52:27.735 INFO  SbtMain:33 |  timeout: 30 seconds
sbtRunner 11:52:27.735 INFO  SbtMain:34 |  isProduction: false
sbtRunner 11:52:27.767 INFO  c.o.s.s.EnsimeActor:165 | Request ensime info from sbtRunner [TELL!]
sbtRunner 11:52:27.832 INFO  c.o.s.s.Sbt(SbtRunner):62 | Starting sbt process
sbtRunner 11:52:31.270 INFO  SbtOutputIO(SbtRunner):104 | [info] Loading global plugins from /home/gui/.sbt/0.13/plugins
sbtRunner 11:52:34.312 INFO  SbtOutputIO(SbtRunner):104 | [info] Loading project definition from /tmp/scastie399751871995553405/project
sbtRunner 11:52:38.330 INFO  SbtOutputIO(SbtRunner):104 | [info] Set current project to scastie399751871995553405 (in build file:/tmp/scastie399751871995553405/)
sbtRunner 11:52:38.509 INFO  SbtOutputIO(SbtRunner):104 | QEbbzADKqP
sbtRunner 11:52:38.510 INFO  c.o.s.s.SbtRunner:135 | Generating ensime config file
sbtRunner 11:52:38.589 INFO  SbtOutputIO(SbtRunner):104 | [error] Not a valid command: ensimeConfig
sbtRunner 11:52:38.590 INFO  SbtOutputIO(SbtRunner):104 | [error] Not a valid project ID: ensimeConfig
sbtRunner 11:52:38.590 INFO  SbtOutputIO(SbtRunner):104 | [error] Expected ':' (if selecting a configuration)
sbtRunner 11:52:38.591 INFO  SbtOutputIO(SbtRunner):104 | [error] Not a valid key: ensimeConfig
sbtRunner 11:52:38.591 INFO  SbtOutputIO(SbtRunner):104 | [error] ensimeConfig
sbtRunner 11:52:38.591 INFO  SbtOutputIO(SbtRunner):104 | [error]             ^
sbtRunner 11:52:38.603 INFO  SbtOutputIO(SbtRunner):104 | QEbbzADKqP
sbtRunner 11:52:38.604 INFO  c.o.s.s.EnsimeActor:250 | Got MkEnsimeConfigResponse
sbtRunner 11:52:38.605 INFO  c.o.s.s.EnsimeActor:176 | Form classpath using .ensime file
sbtRunner [ERROR] [07/12/2017 11:52:38.640] [SbtRemote-akka.actor.default-dispatcher-5] [akka://SbtRemote/user/SbtActor/EnsimeActor] /tmp/scastie399751871995553405/.ensime (No such file or directory)
sbtRunner java.io.FileNotFoundException: /tmp/scastie399751871995553405/.ensime (No such file or directory)
sbtRunner       at java.io.FileInputStream.open0(Native Method)
sbtRunner       at java.io.FileInputStream.open(FileInputStream.java:195)
sbtRunner       at java.io.FileInputStream.<init>(FileInputStream.java:138)
sbtRunner       at scala.io.Source$.fromFile(Source.scala:91)
sbtRunner       at scala.io.Source$.fromFile(Source.scala:76)
sbtRunner       at com.olegych.scastie.sbt.EnsimeActor.com$olegych$scastie$sbt$EnsimeActor$$startEnsimeServer(EnsimeActor.scala:177)
sbtRunner       at com.olegych.scastie.sbt.EnsimeActor$$anonfun$receive$1.applyOrElse(EnsimeActor.scala:251)
sbtRunner       at akka.actor.Actor.aroundReceive(Actor.scala:513)
sbtRunner       at akka.actor.Actor.aroundReceive$(Actor.scala:511)
sbtRunner       at com.olegych.scastie.sbt.EnsimeActor.aroundReceive(EnsimeActor.scala:37)
sbtRunner       at akka.actor.ActorCell.receiveMessage(ActorCell.scala:527)
sbtRunner       at akka.actor.ActorCell.invoke(ActorCell.scala:496)
sbtRunner       at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:257)
sbtRunner       at akka.dispatch.Mailbox.run(Mailbox.scala:224)
sbtRunner       at akka.dispatch.Mailbox.exec(Mailbox.scala:234)
sbtRunner       at akka.dispatch.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
sbtRunner       at akka.dispatch.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
sbtRunner       at akka.dispatch.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
sbtRunner       at akka.dispatch.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
sbtRunner 
sbtRunner 11:52:38.679 INFO  c.o.s.s.EnsimeActor:165 | Request ensime info from sbtRunner [TELL!]
sbtRunner 11:52:38.680 INFO  c.o.s.s.SbtRunner:135 | Generating ensime config file
sbtRunner 11:52:38.751 INFO  SbtOutputIO(SbtRunner):104 | [error] Not a valid command: ensimeConfig
sbtRunner 11:52:38.751 INFO  SbtOutputIO(SbtRunner):104 | [error] Not a valid project ID: ensimeConfig
sbtRunner 11:52:38.752 INFO  SbtOutputIO(SbtRunner):104 | [error] Expected ':' (if selecting a configuration)
sbtRunner 11:52:38.752 INFO  SbtOutputIO(SbtRunner):104 | [error] Not a valid key: ensimeConfig
sbtRunner 11:52:38.753 INFO  SbtOutputIO(SbtRunner):104 | [error] ensimeConfig
sbtRunner 11:52:38.754 INFO  SbtOutputIO(SbtRunner):104 | [error]             ^
sbtRunner 11:52:38.763 INFO  SbtOutputIO(SbtRunner):104 | QEbbzADKqP
sbtRunner 11:52:38.764 INFO  c.o.s.s.EnsimeActor:250 | Got MkEnsimeConfigResponse
sbtRunner 11:52:38.768 INFO  c.o.s.s.EnsimeActor:176 | Form classpath using .ensime file
```

on an infinite loop.

It look like the sbt-ensime plugin is no enabled: https://github.com/scalacenter/scastie/pull/261#discussion_r126910974

File: build.sbt (hunk 1 above)
Comment (dimart) [inline build.sbt]: Got it

File: sbt-runner/src/main/scala/com.olegych.scastie.sbt/EnsimeActor.scala (hunk 5 above)
Comment (dimart) [inline sbt-runner/src/main/scala/com.olegych.scastie.sbt/EnsimeActor.scala]: Ops, I have the sbt-ensime plugin in my global settings.
Gonna also add it to the plugin.sbt then.

File: sbt-runner/src/main/scala/com.olegych.scastie.sbt/EnsimeActor.scala (hunk 4 above)
Comment (dimart) [inline sbt-runner/src/main/scala/com.olegych.scastie.sbt/EnsimeActor.scala]: right >.<

**Comment by dimart** [comment]
fixed issues in #263

**Pull request closed without merging**